src/ray/stats/metric.cc

```diff
@@ -114,24 +112,20 @@ void Metric::Record(double value, const TagsType &tags) {
   if (StatsConfig::instance().IsStatsDisabled()) {
     return;
   }
-  // NOTE(lingxuan.zlx): Double check for recording performance while
-  // processing in multithread and avoid race since metrics may invoke
-  // record in different threads or code pathes.
+
+  absl::MutexLock lock(&registration_mutex_);
   if (measure_ == nullptr) {
```

Comment:
This will have some performance impact. I'm skeptical that it would be significant, but can we at least check the commit that checked this in to see if there were benchmarks included?

Reply:
I don't understand what's the performance impact, in almost all cases it's just a mutex acquisition and release.

Reply:
previously we had no mutex acquisition after the registration path, now we have a shared mutex acquisition on every call to record a metric (one mutex across all metrics). if there are many metrics on many threads, this could cause lock contention.

like I said, I agree that it's unlikely to have meaningful performance impact, but the code may have been added for a reason...

Reply:
This is my reasonable for performance: https://github.com/ray-project/ray/pull/51073#issuecomment-2702491731

Reply:
> like I said, I agree that it's unlikely to have meaningful performance impact, but the code may have been added for a reason...

I'm not sure what do you mean here, we are all clear current impl is buggy and we shouldn't rely on it.